# Titanic

Neste tutorial, faremos uma análise de dados do naufrágio do Titanic, um dos maiores desastres da história naval. A nossa principal tarefa é verificar se dado uma pessoa, caso ela estivesse no Titanic se ela teria sobrevivido ou não.
De forma bem geral, faremos a preparação dos dados de treinamento e teste, identificando dados inconsistentes ou faltantes, numéricos ou categóricos tornando assim, nosso conjunto de dados mais consistente para uma análise mais precisa.
Primeiramente importamos algumas bibliotecas úteis para este tutorial:

In [329]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier 
from sklearn import model_selection
from sklearn.metrics import accuracy_score

O conjunto de dados de treino e de teste estão disponíveis no site Kaggle no seguinte link https://www.kaggle.com/c/titanic/data, porém, no diretório que este projeto está contido, também estão presentes estes dois conjuntos de dados que possuem extensão csv.

In [349]:
treino = pd.read_csv("train.csv")
teste = pd.read_csv("test.csv")

### Tratando os dados de treino e teste.

In [350]:
treino.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [351]:
teste.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

### Obs:
Note que as colunas Name, Ticket e Cabin têm muito pouca, ou nenhuma relação com o fator sobrevivência. Dessa maneira, retiraremos esses dados para esta análise. Perceba que a coluna Survived está presente apenas no conjunto de dados de treino pois é um conjunto de dados que temos informações de sobreviventes. O conjunto de dados de teste é aquele em que queremos saber se uma pessoa sobreviverá ou não.

In [352]:
df_treino_filter = treino[['PassengerId', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked', 'Survived', ]]

df_teste_filter = teste[['PassengerId', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']]


In [353]:
df_treino_filter.head()

PassengerId  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  Survived
0            1       3    male  22.0      1      0   7.2500        S         0
1            2       1  female  38.0      1      0  71.2833        C         1
2            3       3  female  26.0      0      0   7.9250        S         1
3            4       1  female  35.0      1      0  53.1000        S         1
4            5       3    male  35.0      0      0   8.0500        S         0

In [354]:
df_teste_filter.head()

PassengerId  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0          892       3    male  34.5      0      0   7.8292        Q
1          893       3  female  47.0      1      0   7.0000        S
2          894       2    male  62.0      0      0   9.6875        Q
3          895       3    male  27.0      0      0   8.6625        S
4          896       3  female  22.0      1      1  12.2875        S

* ### Atributo Sex (Sexo)
    * O atributo sexo é um tipo de atributo qualitativo ou categórico, desta maneira, é muito útil que se tente quantificar este tipo de atributo pois atributos qualitativos podem conter algumas imprecisões como por exemplo existir na coluna Sexo por exemplo para masculino sendo: (M, m, masc, masculino, Masculino). Perceba que isso dificultaria bastante o trabalho, pois uma atributo que deveria ser binário não está cumprindo este papel, assim adotaremos a seguinte estratégia: para o sexo masculino(male) usaremos 1 e para o feminino(female) utilizaremos o 0. 

In [355]:
df_treino_filter['Sex'] = df_train_filter['Sex'].astype('category').cat.codes

df_teste_filter['Sex'] = df_train_filter['Sex'].astype('category').cat.codes

/home/computer/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/home/computer/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [356]:
df_treino_filter.head()

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  Survived
0            1       3    1  22.0      1      0   7.2500        S         0
1            2       1    0  38.0      1      0  71.2833        C         1
2            3       3    0  26.0      0      0   7.9250        S         1
3            4       1    0  35.0      1      0  53.1000        S         1
4            5       3    1  35.0      0      0   8.0500        S         0

In [357]:
df_teste_filter.head()

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0          892       3    1  34.5      0      0   7.8292        Q
1          893       3    0  47.0      1      0   7.0000        S
2          894       2    0  62.0      0      0   9.6875        Q
3          895       3    0  27.0      0      0   8.6625        S
4          896       3    1  22.0      1      1  12.2875        S

* ### Atributo Embarked (Portão de embarque)
    * O atributo Embarked é outro tipo de atributo qualitativo ou categórico, desta maneira, adotaremos a seguinte estratégia: Portão C = 1 , Q = 2 e S = 3 .  

In [358]:
df_treino_filter['Embarked'] = df_treino_filter['Embarked'].astype('category').cat.codes

df_teste_filter['Embarked'] = df_teste_filter['Embarked'].astype('category').cat.codes


/home/computer/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/home/computer/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [359]:
df_treino_filter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
PassengerId    891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null int8
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
Embarked       891 non-null int8
Survived       891 non-null int64
dtypes: float64(2), int64(5), int8(2)
memory usage: 50.5 KB


In [360]:
df_teste_filter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 8 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null int8
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Fare           417 non-null float64
Embarked       418 non-null int8
dtypes: float64(2), int64(4), int8(2)
memory usage: 20.5 KB


* ### Atributo Age (Idade)
    * Como podemos verificar no comando utilizado acima, das 891 linhas do conjunto de treino, apenas 714 estão preenchidos com idade e no conjunto de teste das 418, apenas 332 estão preenchidas, dessa maneira, devemos preencher estes campos nulos com algum valor. Neste tutorial, usaremos a média das idades, porém, existem outras estratégias para a escolha de uma idade a ser preenchida.

In [361]:
media_idade_treino = df_treino_filter['Age'].mean()
print("Media de idade (conjunto de treino): ",media_idade_treino)

media_idade_teste = df_teste_filter['Age'].mean()
print("Media de idade (conjunto de teste): ",media_idade_teste)

Media de idade (conjunto de treino):  29.69911764705882
Media de idade (conjunto de teste):  30.272590361445783


In [362]:
## Preencher camos de idade vazios com a média das idades
df_treino_filter['Age'].fillna(media_idade_treino, inplace=True)
df_teste_filter['Age'].fillna(media_idade_teste, inplace=True)

/home/computer/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:3549: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [363]:
X.isnull().any()

Pclass      False
Sex         False
Age         False
SibSp       False
Parch       False
Fare        False
Embarked    False
dtype: bool

In [364]:
X = df_train_filter.drop(['PassengerId', 'Survived'], axis=1)
X_teste = df_teste_filter.drop(['PassengerId'], axis=1)
y = df_train_filter['Survived']

### Usando nosso classificador

Utilizaremos uma árvore de decisão para classificar nosso conjunto de dados.

In [365]:
mapa = {0:'NÃO SOBREVIVENTE',1:'SOBREVIVENTE'}

classificador = DecisionTreeClassifier()

classificador = classificador.fit(X,y)

predicao = classificador.predict(X_teste.loc[20])
print(mapa[predicao[0]])

SOBREVIVENTE


/home/computer/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


In [366]:
result = model_selection.cross_val_score(classificador, X, y, cv=10)
result

array([ 0.7       ,  0.77777778,  0.70786517,  0.80898876,  0.82022472,
        0.74157303,  0.82022472,  0.73033708,  0.82022472,  0.81818182])

In [367]:
np.average(result)

0.77453977982067868